### PR TITLE
[2.0 change] Hard-fail on invalid write types in core channels.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

There is relatively little benefit in allowing the user to send invalid
data types on a given channel. Currently we tolerate this by simply
failing the promise, but this is not really ideal: if the user never sets
a promise here they never even find out that we bailed on sending their
data. Given that we attempt to be type safe everywhere else, we should
just blow up immediately if the user attempts to write an invalid data
type.

Modifications:

Removed the nice gentle type casts, replaced them with explody ones.

Result:

Faster failure in error cases.

### CODE REVIEW

File: Sources/NIO/SocketChannel.swift [hunk 1]
```diff
@@ -290,11 +290,7 @@ final class SocketChannel: BaseSocketChannel<Socket> {
             return
         }
 
-        guard let data = data.tryAsIOData() else {
-            promise?.fail(error: ChannelError.writeDataUnsupported)
```
Comment (normanmaurer) [inline Sources/NIO/SocketChannel.swift]: I think we should also remove `ChannelError.writeDataUnsupported` completely ?

Comment (Lukasa) [reply Sources/NIO/SocketChannel.swift]: ACK, done.

**Comment by weissi** [comment]
@Lukasa this conflicts now

**Comment by Lukasa** [comment]
Updated.

**Comment by weissi** [comment]
@swift-nio-bot test this please

**Comment by weissi** [comment]
@normanmaurer are you happy with this?

**Comment by normanmaurer** [comment]
I am not strong on this honestly... I am fine either way.

**Comment by weissi** [comment]
@Lukasa for you to decide

**Pull request merged**